### ISSUE TITLE
how do I run parallel mode via API?

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#312]
*Originally reported by* **Anonymous**

----------------------------------------

My flow goes like this:

```
  cov = coverage.coverage(data_suffix=True)
  cov.start()
  suite = unittest.defaultTestLoader.discover(...)
  suite = ConcurrentTestSuite(suite, fork_for_tests(4))  # os.fork 4 subprocesses
  unittest.TextTestRunner().run(suite)
  cov.stop()
  cov.combine()
  cov.save()
```
My suite does a lot of forks (via http://coreygoldberg.blogspot.com/2013/06/python-concurrencytest-running.html).   

I thought setting data_suffix=True would force the parallel runs, yet I get different coverage results depending if I os.fork the runs or not.   

  

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/312
- This issue had attachments: concurrent-example.tgz.  See the [original issue](https://bitbucket.org/ned/coveragepy/issue/312) for details.

**Comment by nedbat** [comment]
Can you provide a small sample project that demonstrates the problem? (Even a large project that demonstrates the problem would be good.)

**Comment by nedbat** [comment]
*Original comment by* **Albert Ting (Bitbucket: [altlist](https://bitbucket.org/altlist), GitHub: Unknown)**

----------------------------------------

Attached is an example.  

```
#!bash

$ run_tests --parallel=1    -> 100% coverage
$ run_tests --parallel=10  ->  35% coverage
```

**Comment by nedbat** [comment]
*Original comment by* **Albert Ting (Bitbucket: [altlist](https://bitbucket.org/altlist), GitHub: Unknown)**

----------------------------------------

Hi, are there any updates?

**Comment by nedbat** [comment]
Thanks, I've reproduced it, and will dig in soon to understand what's going wrong.

**Comment by nedbat** [comment]
@altlist This seems to be a duplicate of #310, os.fork and os._exit.

**Comment by nedbat** [comment]
Duplicate of #310.

**Comment by nedbat** [comment]
In commit 44a309f9c8aeb140b64aa45453e0595a85c5c9f9, I added a new configuration setting, `[run] patch` that you can use to fix `os._exit()`.

Add this to you .coveragerc (or other configuration file), and the coverage data will be saved when `os._exit` is used:
```ini
[run]
patch = _exit
```

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.10.0](https://pypi.org/project/coverage/7.10.0).
